internal/peer/node/start.go

```diff
@@ -948,7 +950,7 @@ func registerDiscoveryService(
 	polMgr policies.ChannelPolicyManagerGetter,
 	metadataProvider *lifecycle.MetadataProvider,
 	gossipService *gossipservice.GossipService,
-) {
+) gateway.DiscoveryService {
```

Comment:
We should not be using a foreign interface type as a return value from what is logically a constructor.

Reply:
We have a common idiom in go: "accept interfaces, return structs". In this case, we should be returning the concrete type from the discovery package. That said, the discovery package is not idiomatic go and has exported a constructor that returns an unexported struct. This means that you, as the consumer, can't assign it to anything other than an interface you define.

The code in the discovery package should really be exporting `service` as `Service`.  The exported fields remain inaccessible but this code can now use the appropriate concrete type.

Later, when you go to register the implementation, your gateway interface would be used to ensure that what the caller provides satisfies your contract.

With that...
1. we should have a change before this that exports `service` as `Service` from the discovery layer
2. we should avoid explicitly binding to gateway.DiscoveryService interface within this package if possible

Finally, a consistent problem within fabric is that we define our interfaces before the implementation. This is backwards for go. Interfaces should be defined when we need to enforce a contract on a type we accept instead of defining the contract we want to implement.

Reply:
Just to follow-up on the comment about interfaces, please see https://github.com/golang/go/wiki/CodeReviewComments#interfaces (referenced in our style guide) for details.